DOL.WHD.Section14c.Business/Services/AttachmentService.cs

```diff
@@ -16,27 +18,33 @@ public class AttachmentService : IAttachmentService
         private readonly IFileRepository _fileRepository;
         private readonly IAttachmentRepository _attachmentRepository;
         private bool Disposed = false;
+        private readonly string FileEncryptKey = "@WSX3edc$RFV5tgb^YHN7ujm";
```

Comment:
Crypto secrets must never be in source code.  This needs to come from an external source - something that is not versioned or distributed with the source code.  It could be an environment variable, or it could go into a config file that is ***NOT*** versioned in git (e.g., a sample of the config file with no encryption secret set could be added to the repo, and then the file could be added to `.gitignore` so that updates to the file would not be versioned).  The ideal case is to use a secrets manager with an API, but that's maybe a big reach this late in development.

Questions for @mmurthydo @binwang89 and DOL IT:

- How sensitive are the files being encrypted?  That will play a large role in understanding how well-protected the crypto secret must be.
- Who are we trying to prevent from reading these files by encrypting them?
  - Can OS-level full-disk encryption satisfy this requirement?
- How long must these files be accessible?  Are they irrelevant after a year, or is there a potential to need them several years from now (e.g., legal action or something)?
- Is there a plan for how to rotate the secret if it's leaked?  Will all the files be decrypted with the old secret and re-encrypted with the new one?  Is there a process outlined for how to do that?

Encrypting every file with the same key also increases the risk of the key being broken.  I think a better practice in this case is to generate a unique, random key for each file and then encrypt *that* key with the "master" key - that way the total amount of data encrypted with the master key is relatively small (a few dozen bytes per file).  It's a little more complicated to manage, but it enhances security a good bit.  (However, depending on how high the security needs, this may be overkill.)

Reply:
Per discussion with Madhu, I added an additional column in the attachment table to store the encryption key. The encryption key is unique and randomly generated for each file. The encryption algorithm is based on NIST's Cryptographic Standards and Guidelines, AES algorithm is used with cryptographic keys of 256 bits to encrypt and decrypt data in blocks of 128 bits.